Ignore `#!/...` shebang on first line of code files

### Check for existing issues

- [X] Completed

### Describe the feature

Currently, vale process a `#!/...` shebang on the first line of a code file as a comment.  For example, when run on a Python script with a first line of `#!/usr/bin/env python3`, vale produces the following error:

```
 1:12   error       Use 'environment' rather than   RedHat.TermsErrors      
                    'env'.                                                   
```

In terms of being syntax aware, a shebang line should not be considered a comment.  Add an exception to the logic that identifies comments to exclude comments starting with `!/` if that comment is on the first line.

--- a/internal/lint/code/comments.go
+++ b/internal/lint/code/comments.go
@@ -149,5 +149,26 @@ func GetComments(source []byte, lang *Language) ([]Comment, error) {
 		})
 	}
 
-	return coalesce(comments), nil
+	return coalesce(dropShebang(comments)), nil
+}
+
+// dropShebang removes a leading `#!` line.
+//
+// A shebang names the interpreter a script runs under. Languages that use `#`
+// for comments give tree-sitter no way to tell the two apart, so it arrives
+// here as a comment and is linted as prose -- a path like `/usr/bin/env` then
+// draws alerts about words the author never wrote. See #631.
+//
+// Only the very first line of the file qualifies, which is also all the kernel
+// accepts, so a `#!` written anywhere else stays a comment.
+func dropShebang(comments []Comment) []Comment {
+	for i, c := range comments {
+		if c.Line != 1 || c.Offset != 0 {
+			continue
+		} else if !strings.HasPrefix(c.Source, "#!") {
+			continue
+		}
+		return append(comments[:i:i], comments[i+1:]...)
+	}
+	return comments
 }

--- a/testdata/e2e/lint.yaml
+++ b/testdata/e2e/lint.yaml
@@ -148,6 +148,30 @@ cases:
       test.py:35:8:vale.Annotations:'NOTE' left in text
       test.py:37:5:vale.Annotations:'TODO' left in text
 
+  - name: python/shebang
+    about: "#631 -- a shebang names an interpreter, so it isn't a comment"
+    files:
+      .vale.ini: |
+        StylesPath = styles
+        MinAlertLevel = suggestion
+
+        [*.py]
+        BasedOnStyles = Test
+      styles/Test/Env.yml: |
+        extends: existence
+        message: "Use 'environment' rather than 'env'."
+        level: error
+        tokens:
+          - env
+      test.py: |
+        #!/usr/bin/env python3
+        # Read the env before starting.
+        print("hello")
+    args: test.py
+    exit: 1
+    want: |
+      test.py:2:12:Test.Env:Use 'environment' rather than 'env'.
+
   - name: cpp
     args: test.cc
     exit: 0

